src/librustc_borrowck/borrowck/gather_loans/move_error.rs

```diff
@@ -153,20 +153,26 @@ fn report_cannot_move_out_of<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>,
         }
 
         Categorization::Interior(ref b, mc::InteriorElement(ik)) => {
-            match (&b.ty.sty, ik) {
-                (&ty::TySlice(..), _) |
-                (_, Kind::Index) => {
-                    let mut err = struct_span_err!(bccx, move_from.span, E0508,
-                                                   "cannot move out of type `{}`, \
-                                                    a non-copy array",
-                                                   b.ty);
-                    err.span_label(move_from.span, "cannot move out of here");
-                    err
-                }
-                (_, Kind::Pattern) => {
-                    span_bug!(move_from.span, "this path should not cause illegal move");
-                }
-            }
+            let type_name = match &b.ty.sty {
```

Comment:
Can you maintain the `bug!` on `b.ty.sty == ty::TyArray` and `ik == Kind::Pattern`?